NE-1092: Add proxy protocol support for IBMCloud loadbalancers

SzucsAti opened the pull request:
On IBMCloud users can set a loadbalancer annotation to enable proxy protocol on the loadbalancer: 
`service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features: "proxy-protocol"`
https://cloud.ibm.com/docs/containers?topic=containers-vpc-lbaas

With this PR users can configure proxy protocol on the ingresscontroller custom resource. Cluster-ingress-operator will read the value of this field and configure the loadbalancer accordingly.
```
  endpointPublishingStrategy:
    loadBalancer:
      providerParameters:
        type: IBM
        ibm:
          protocol: PROXY
      scope: External
    type: LoadBalancerService
```

Currently using `openshift/api` from a fork that contains the related api changes. Will update the dependency after the api changes get merged to master. Related api PR: https://github.com/openshift/api/pull/1236

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -93,6 +93,12 @@ const (
 	// IP as private.
 	iksLBScopePrivate = "private"
 
+	// iksLBEnableFeatures is the annotation used on a service to enable features
```
```suggestion
	// iksLBEnableFeaturesAnnotation is the annotation used on a service to enable features
```

Miciah commented on manifests/00-custom-resource-definition.yaml:
```diff
@@ -292,10 +292,41 @@ spec:
+                                  at https://cloud.ibm.com/docs/containers?topic=containers-vpc-lbaas"
+                                properties:
+                                  proxyProtocol:
+                                    description: "Enable PROXY protocol on the loadbalancer.
+                                      The load balancer passes client connection information,
+                                      including the client IP address, the proxy server
+                                      IP address, and both port numbers, in request
+                                      headers to your back-end app. \n See \"service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features:
+                                      \"proxy-protocol\"\" at https://cloud.ibm.com/docs/containers?topic=containers-vpc-lbaas"
```
This description should mention that PROXY protocol cannot be toggled on an ingresscontroller after it has been created; toggling will require deleting and recreating the ingresscontroller.  

If you actually want toggling PROXY protocol to be possible without deleting and recreating the ingresscontroller, then `setDefaultPublishingStrategy` needs some logic to check for updates to `spec.endpointPublishingStrategy.loadBalancer.providerParameters.ibm.enableFeatures.proxyProtocol`.  However, toggling the setting would most likely cause brief disruption as the LB and router deployment would be updated asynchronously, and the router drops traffic during any period of time during which the configuration is inconsistent (that is, the LB is sending PROXY and HAProxy is expecting TCP or vice versa).  

Would it be feasible to use the `IngressControllerProtocol` type for this field?  That would make the `spec.endpointPublishingStrategy.loadBalancer.providerParameters.ibm.enableFeatures.proxyProtocol` API field more consistent with the `spec.endpointPublishingStrategy.hostNetwork.protocol`, `spec.endpointPublishingStrategy.private.protocol`, and `spec.endpointPublishingStrategy.nodePort.protocol` fields.  You could also re-use some of the wording from the godoc for those fields.

SzucsAti replied:
- Updated the variable in `openshift/api`: Renamed it to`spec.endpointPublishingStrategy.loadBalancer.providerParameters.ibm.protocol` and changed the type from `bool` to `IngressControllerProtocol`. 
- Added logic to `setDefaultPublishingStrategy` so routers get updated after setting `spec.endpointPublishingStrategy.loadBalancer.providerParameters.ibm.protocol`.

One thing that doesn't seem to work is automatic loadbalancer updates. On IBMCloud the proxy protocol annotation has to be added to the loadbalancer during creation time for it to work. This means that the ingress operator would have to delete the loadbalancer and recreate it. We can also just document and ask the users to delete the loadbalancer service manually and ingress operator will recreate it for them.

Miciah reviewed (commented):
Is there a Jira epic for this feature?

attiss commented:
Thanks for the review @Miciah!

>Is there a Jira epic for this feature?

We don't have an epic for that (yet). Please let us know how we could create one.

SzucsAti commented:
Thank you @Miciah for the review! Updated the PR based on your comments.

On IBMCloud we currently support ROKS 4.7-4.11 versions. Would it be possible to backport/cherry-pick this feature for these versions?

We currently don't have a Jira issue for this feature, I think we would need someone to create one, as we most likely don't have the required permissions(?). Do you know who we can contact to help us?

SzucsAti commented:
/retest

SzucsAti commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -93,6 +93,12 @@ const (
 
+	// iksLBEnableFeaturesAnnotation is the annotation used on a service to enable features
+	// on the load balancer.
+	iksLBEnableFeaturesAnnotation = "service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features"
```
Added `iksLBEnableFeaturesAnnotation` to `managedLoadBalancerServiceAnnotations`. Test case seems to work now.

jeffnowicki commented:
/retest-required

candita commented:
/assign @Miciah 
/assign @gcs278

jeffnowicki commented:
/retest-required

lihongan commented:
/retest-required

SzucsAti commented:
/retest-required

lihongan commented:
tested with pre-merge validation and passed

```
$ oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.ci.test-2023-02-07-013419-ci-ln-k91zc62-latest   True        False         139m    Cluster version is 4.13.0-0.ci.test-2023-02-07-013419-ci-ln-k91zc62-latest

### after updating the ingresscontroller/default and got 
$ oc -n openshift-ingress-operator get ingresscontroller/default -oyaml
<---snip-->
spec:
  endpointPublishingStrategy:
    loadBalancer:
      dnsManagementPolicy: Managed
      providerParameters:
        ibm:
          protocol: PROXY
        type: IBM
      scope: External
    type: LoadBalancerService
<------>
status:
  domain: apps.hongli-ibm.ibmcloud.qe.devcluster.openshift.com
  endpointPublishingStrategy:
    loadBalancer:
      dnsManagementPolicy: Managed
      providerParameters:
        ibm:
          protocol: PROXY
        type: IBM
      scope: External
    type: LoadBalancerService
<------>

### ensure router pods use PROXY protocol
$ oc -n openshift-ingress get deployment/router-default -oyaml
<---snip--->
        - name: ROUTER_USE_PROXY_PROTOCOL
          value: "true"

### ensure the proxy-protocol annotation is added to the LB service 
$ oc -n openshift-ingress get svc/router-default -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features: proxy-protocol
    service.kubernetes.io/ibm-load-balancer-cloud-provider-ip-type: public
<---snip--->

### ensure routes and ingress operator work well
$ curl https://canary-openshift-ingress-canary.apps.hongli-ibm.ibmcloud.qe.devcluster.openshift.com -k
Healthcheck requested
 
$ oc get co/ingress
NAME      VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
ingress   4.13.0-0.ci.test-2023-02-07-013419-ci-ln-k91zc62-latest   True        False         False      153m    
```

lihongan commented:
/label qe-approved

Miciah commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -561,8 +561,29 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
+				// The only provider parameter that is supported
+				// for IBM is the Protocol parameter.
+				var specProtocol, statusProtocol operatorv1.IngressControllerProtocol
+				if specLB.ProviderParameters != nil && specLB.ProviderParameters.IBM != nil {
+					specProtocol = specLB.ProviderParameters.IBM.Protocol
+				}
```
Rather than checking `specLB.ProviderParameters != nil && specLB.ProviderParameters.IBM` here, please add a case in `setDefaultProviderParameters` to initialize these fields like we have for `operatorv1.GCPLoadBalancerProvider`.

SzucsAti replied:
Moved specLB initialization to `setDefaultProviderParameters`.

ahardin-rh commented:
/label docs-approved

CFields651 commented:
/label px-approved

Miciah commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -561,8 +561,29 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
+				if statusLB.ProviderParameters != nil && statusLB.ProviderParameters.IBM != nil {
+					statusProtocol = statusLB.ProviderParameters.IBM.Protocol
+				}
+				if specProtocol != statusProtocol {
+					if statusLB.ProviderParameters == nil {
+						statusLB.ProviderParameters = &operatorv1.ProviderLoadBalancerParameters{
+							Type: operatorv1.IBMLoadBalancerProvider,
+						}
+					}
+					if statusLB.ProviderParameters.IBM == nil {
+						statusLB.ProviderParameters.IBM = &operatorv1.IBMLoadBalancerParameters{}
+					}
+					statusLB.ProviderParameters.IBM.Protocol = specProtocol
+					changed = true
+				}
```
Follow the style we use in the other cases in order to reduce nesting:
```suggestion
				if specProtocol != statusProtocol {
					if statusLB.ProviderParameters == nil {
						statusLB.ProviderParameters = &operatorv1.ProviderLoadBalancerParameters{}
					}
					if len(statusLB.ProviderParameters.Type) == 0 {
							statusLB.ProviderParameters.Type = operatorv1.IBMLoadBalancerProvider
					}
					if statusLB.ProviderParameters.IBM == nil {
						statusLB.ProviderParameters.IBM = &operatorv1.IBMLoadBalancerParameters{}
					}
					statusLB.ProviderParameters.IBM.Protocol = specProtocol
					changed = true
				}
```

gcs278 reviewed (commented):
Reviewed - I don't have any comments other than what Miciah suggested. Otherwise, looks good to me.

SzucsAti commented:
Thank you for the reviews. Aside from addressing your comments, I added a few tcs to `TestSetDefaultPublishingStrategyHandlesUpdates`, which I think was missing.

Before merging this PR, I'll also need to remove the `replace` directive from `go.mod`, which can be done once the related`openshift/api` PR is merged.

SzucsAti commented:
Updated `go.mod` file to use the latest `openshift/api` and removed the replace directive.

Miciah commented:
Thanks!
/approve
/lgtm
/hold
@gcs278, did you want to give this another look?

gcs278 commented:
/lgtm
/unhold

Thanks!

lihongan commented:
@SzucsAti I found a new platform also uses IBM load balancer, could you help confirm if we need another PR to support the `PowerVs` ?
```
$ oc get infrastructures.config.openshift.io cluster -oyaml | yq .status
apiServerInternalURI: https://api-int.rdr-ipi2-<hide>:6443
apiServerURL: https://api.rdr-ipi2-<hide>:6443
controlPlaneTopology: HighlyAvailable
cpuPartitioning: None
etcdDiscoveryDomain: ""
infrastructureName: rdr-ipi2-xxxxx
infrastructureTopology: HighlyAvailable
platform: PowerVS
platformStatus:
  powervs:
    cisInstanceCRN: 'crn:v1:bluemix:public:internet-svcs:global:<hide>::'
    region: wdc
    resourceGroup: ipi-resource-group
    zone: wdcxx
  type: PowerVS

$ oc -n openshift-ingress get svc router-default -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.kubernetes.io/ibm-load-balancer-cloud-provider-ip-type: public
  creationTimestamp: "2024-04-08T11:01:49Z"
  finalizers:
  - service.kubernetes.io/load-balancer-cleanup
<......>
```
